### DISCUSSION TITLE
Description  for confidence image values

### DISCUSSION THREAD

**Discussion by Astronomia-maker** [ifm/ifm3d#394] [Q&A]
Hello,
I have O3R 225 camera and installled ifm3dpy==1.1.1 for my python environment. I get array for confidence image, but values in arrays does not match [this description](https://ifm3d.com/documentation/DeviceConfiguration/ImagesDescription/confidenceImage.html?highlight=confidence). 
This is the output for arrays that i am getting. 
`
 [[  577   577   577 ...   513   513   577]
 [  577   577   577 ...   513   513   577]
 [  577   577    97 ...    41   513   513]
 ...
 [  513   513 16417 ...  4129   513   513]
 [  513   513   513 ...   513   513   513]
 [  513   513   513 ...   513   513   513]]
` 
So where can i find the description for the confidence image values?

**Comment by lola-masson** [comment]
Hi @Astronomia-maker, 
The values in the confidence image can be combinations of multiple values. For instance:

577 = 512 + 64 + 1 -> Edge or isolated pixel (512) + invalid range (64) + invalid (1)  
513 = 512 + 1 -> Edge or isolated pixel + invalid  
97 = 64 + 32 + 1 -> Invalid range + Exposure index (32) + invalid

**Reply by Astronomia-maker** [reply]
Hi @lola-masson, thank you for the reply. is there any way that i can have all confidence image value without addition?

**Reply by lola-masson** [reply]
Unfortunately, this is the only way the confidence image is streamed from the device.

**Reply by BigBoot** [reply]
Hello @Astronomia-maker,

could you please clarify what value you're interested in exactly?

The values in the confidence image are 16bit bitflags, that means each bit in the value represents whether a flag is set or not.
To check if a pixel has a specific flag set you can use the `bitwise and` operator like this:

```python
if img[42][42]&1024: 
    print("Pixel 42 42 is UNPLAUSIBLE")
```

You can also apply the `bitwise and` to the whole image like this:
```python
unplausible = img&1024
```
which would result in an image where each value is either `1024` if the flag is set  or `0` if it isn't.

**Reply by Astronomia-maker** [reply]
Hi @BigBoot, Thank you for the explanation. Actually i want to eliminate all pixel which have crosstalk error, some reflected pixel from the object if there is a box then there are some flying pixel around the box (not every time i have). So the confidence image value could be ( [document reference:-](https://ifm3d.com/documentation/DeviceConfiguration/ImagesDescription/confidenceImage.html?highlight=confidence)32768,4096,16384,4,2) as far my knowledge.

**Reply by BigBoot** [reply]
In that case you could use something like the following:
```python
flags = (32768|16384|4096|4|2)
mask = np.not_equal(confidence&flags, 0)
```
which results in a boolean mask like this:
```
[[ True  True  True ...  True  True  True]
 [ True  True  True ...  True  True  True]
 [False  True  True ...  True False  True]
 ...
 [ True  True  True ...  True  True  True]
 [ True  True  True ...  True  True  True]
 [ True  True  True ...  True  True  True]]
```

which can then be used to mask your initial image:
```python
masked = np.ma.masked_array(img, mask=mask)
```

**Reply by Astronomia-maker** [reply]
Hi @BigBoot, Thank you for the suggestion and explanation. I will check out this suggested solution.
